synapse/rest/admin/users.py

```diff
@@ -0,0 +1,76 @@
+# -*- coding: utf-8 -*-
+# Copyright 2019 The Matrix.org Foundation C.I.C.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import re
+
+from twisted.internet import defer
+
+from synapse.api.errors import SynapseError
+from synapse.http.servlet import (
+    RestServlet,
+    assert_params_in_dict,
+    parse_json_object_from_request,
+)
+from synapse.rest.admin import assert_requester_is_admin
+from synapse.types import UserID
+
+
+class UserAdminServlet(RestServlet):
+    """
+    Set whether or not a user is a server administrator.
+
+    Note that only local users can be server administrators, and that an
+    administrator may not demote themselves.
+
+    Only server administrators can use this API.
+
+    Example:
+        PUT /_synapse/admin/v1/users/@reivilibre:librepush.net/admin
+        {
+            "admin": true
+        }
+    """
+
+    PATTERNS = (re.compile("^/_synapse/admin/v1/users/(?P<user_id>@[^/]*)/admin$"),)
```

Comment:
for better or worse, userids are allowed to contain slashes (https://matrix.org/docs/spec/appendices#user-identifiers). Any idea how this will behave?

Reply:
It seems to handle them correctly: (I created the user @/dev/null:sawa on a local test server):

```
 ~  curl -X PUT -H 'Authorization: Bearer MDAxMmxvY2F0aW9uIHNhd2EKMDAxM2lkZW50aWZpZXIga2V5CjAwMTBjaWQgZ2VuID0gMQowMDFlY2lkIHVzZXJfaWQgPSBAc2F3YTM6c2F3YQowMDE2Y2lkIHR5cGUgPSBhY2Nlc3MKMDAyMWNpZCBub25jZSA9IDV0Wi5qPUp2bDIxb2FkfnYKMDAyZnNpZ25hdHVyZSBYzm92tHOvLu6w-el6662J8ikDjmf4nR7yowmzM0fDSQo' http://sawa:8008/_synapse/admin/v1/users/@%2Fdev%2Fnull:sawa/admin -d '{"admin": false}'
null
 ~  curl -X PUT -H 'Authorization: Bearer MDAxMmxvY2F0aW9uIHNhd2EKMDAxM2lkZW50aWZpZXIga2V5CjAwMTBjaWQgZ2VuID0gMQowMDFlY2lkIHVzZXJfaWQgPSBAc2F3YTM6c2F3YQowMDE2Y2lkIHR5cGUgPSBhY2Nlc3MKMDAyMWNpZCBub25jZSA9IDV0Wi5qPUp2bDIxb2FkfnYKMDAyZnNpZ25hdHVyZSBYzm92tHOvLu6w-el6662J8ikDjmf4nR7yowmzM0fDSQo' http://sawa:8008/_synapse/admin/v1/users/@%2Fdev%2Fnulls:sawa/admin -d '{"admin": false}'
{
    "errcode": "M_UNKNOWN",
    "error": "No row found (users)"
}
```

However, one thing I need to sort out is this `null` response on success. I obviously got a bit excited and overlooked it before submitting this PR. What should be returned on success? Empty dict?

Reply:
jolly good. looks like you resolved the `null` thing too.